core/constraints/validation.go

```diff
@@ -222,13 +226,42 @@ func (v *validator) checkInVocab(attributeName string, attributeValue interface{
 	if !ok {
 		return nil
 	}
+
+	coercedAttributeValue := coerce(attributeValue)
 	for _, validValue := range validValues {
-		if coerce(validValue) == coerce(attributeValue) {
+		if coerce(validValue) == coercedAttributeValue {
 			return nil
 		}
 	}
+
+	attributeValueStr := fmt.Sprint(attributeValue)
+	validValuesStrLst := make([]string, 0)
+	validValuesStrMap := make(map[string]struct{})
+
+	// Collect unique constraint values
+	for _, value := range validValues {
+		valueStr := fmt.Sprint(value)
+		if _, exists := validValuesStrMap[valueStr]; !exists {
+			validValuesStrMap[valueStr] = struct{}{}
+			validValuesStrLst = append(validValuesStrLst, valueStr)
+		}
+	}
+
+	// Sort constraint values according to LevenshteinDistance between attributeValue and itself
+	sort.Slice(validValuesStrLst, func(i, j int) bool {
+		d1 := stringcomparator.New(validValuesStrLst[i], attributeValueStr).LevenshteinDistance()
+		d2 := stringcomparator.New(validValuesStrLst[j], attributeValueStr).LevenshteinDistance()
+		return d1 < d2
+	})
+
+	// Return the closest match if there are too many valid values
+	if v.validValuesCountLimit >= 0 && len(validValuesStrLst) > v.validValuesCountLimit {
+		return fmt.Errorf(
+			"invalid constraint value: %v=%v\ndid you mean: %v?", attributeName, attributeValue, validValuesStrLst[0])
```

Comment:
We don't want to surface a user facing error string here. We want to return a typed error - create a new error struct with attributes for the allowed values, the supplied input, and the suggestion; thread that error struct back through to the caller. Then the caller decides how to represent the error. 
Also, the "did you..." messages needs to be "Did you...."

Reply:
Hi Ian, I was wondering do we need the supplied input in the struct since the caller already knows about the input?Also, do you think it might be more appropriate to name the struct ValidatorResponse instead of ValidatorError, since it includes the suggested valid values slice?